### ISSUE TITLE
Lost type name when its length is longer than 255

### ISSUE THREAD

**Issue by al6uiz** [mgholam/fastBinaryJSON#7]
```
            byte[] b = Reflection.Instance.utf8.GetBytes(s);
            _output.WriteByte((byte)b.Length);
            _output.Write(b, 0, b.Length % 256);
```
This code causes incorrect parsing type name when its length is longer than 255.

My type is generic and has long name. 

JIT.MyProject.Shared.KeyValue`2[[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]][], JIT.MyProject.Shared, Version=1.0.6644.39299, Culture=neutral, PublicKeyToken=null

**Comment by mgholam** [comment]
The name token is limited to 255 bytes in length.

**Comment by al6uiz** [comment]
The problem is the name token is used for not only property name but also **typedarray**.
In real case, it usually happens the length of type's name(including namespace) is longer than 255.

Can you consider my opinion?

My suggestion is this.

```
if (length < 0xFF)
    [length(1)] [string bytes(length)]
else
    [0xFF] [legnth(2)] [string bytes(length)]
```
In normal case, it may works same as before and in the long case, it just uses one more extra byte.

**Comment by mgholam** [comment]
Can you give me some sample code to test with?

**Comment by al6uiz** [comment]
Please try this code.

https://gist.github.com/al6uiz/0c5c1fb669834f8e64cd234eb1946585

**Comment by al6uiz** [comment]
And to fix this problem, I modified BJsonSerializer.cs

```
private unsafe void WriteNameInternal(string name)
{
    byte[] b = Reflection.Instance.utf8.GetBytes(name);

    if (b.Length >= 0xFF)
    {
        _output.WriteByte(0xFF);

        ushort length = (ushort)b.Length;
        byte* p = (byte*)&length;
                
        _output.WriteByte(p[0]);
        _output.WriteByte(p[1]);
    }
    else
    {
        _output.WriteByte((byte)b.Length);
    }
    _output.Write(b, 0, b.Length);
}
```

and BJsonParser.cs
```
private string ParseName()
{
    short length = _json[_index++];

    if (length == 0xFF)
    {
        length = Helper.ToInt16(_json, _index);

        _index += 2;
    }


    string s = Reflection.Instance.utf8.GetString(_json, _index, length);
    _index += length;
    return s;
}
```

**Comment by mgholam** [comment]
Fixed now in `v1.4.22 ` with  fewer bytes :)
